### PR TITLE
Jkube addon dev steps and demonstration

### PR DESCRIPTION
**Pull request by mundra-ankur**
Signed-off-by: Ankur Mundra <20452032+mundra-ankur@users.noreply.github.com>

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,7 +1,103 @@
...
+This is an implementation of [Jkube's](https://www.eclipse.org/jkube) `kubernetes-maven-plugin` as a konveyor addon. This addon is a small step towards Application Modernization. It brings your <strong>Java</strong> applications on to Kubernetes by leveraging the tasks required to make your application cloud-native. It provides a tight integration into Maven and benefits from the build configuration already provided. 
 
+This addon focus on two tasks: 
+- Generates container images with flexible and powerful configuration.
+- Generates vanilla Kubernetes descriptors (YAML files).
+
+It leverage the [Kubernetes Maven plugin](https://www.eclipse.org/jkube/docs/kubernetes-maven-plugin)  to generate conatiner image and Kubernetes resources.
+
+## Development
+To browse code - [Open in VSCode](https://open.vscode.dev/konveyor/tackle2-addon-jkube)
```
Comment (djzager) [inline README.md]: Nice. Very useful.

File: README.md [hunk 2]
```diff
@@ -1,7 +1,103 @@
 Tackle2 Addon Jkube
 ===================
-The addon is based on the [Tackle2](https://github.com/konveyor?q=tackle2&type=all&language=&sort=) project.
+The addon is part of the [Tackle2](https://github.com/konveyor?q=tackle2&type=all&language=&sort=) projects.
 
-This addon generate Kubernetes resources from a Java application.
-It uses the [JKube](https://www.eclipse.org/jkube/docs/kubernetes-maven-plugin) Maven plugin to generate the Kubernetes resources.
+This is an implementation of [Jkube's](https://www.eclipse.org/jkube) `kubernetes-maven-plugin` as a konveyor addon. This addon is a small step towards Application Modernization. It brings your <strong>Java</strong> applications on to Kubernetes by leveraging the tasks required to make your application cloud-native. It provides a tight integration into Maven and benefits from the build configuration already provided.
```
Comment (djzager) [inline README.md]: Not perfect but here is a suggestion to improve:

```suggestion
This project exposes [JKube's](https://www.eclipse.org/jkube) `kubernetes-maven-plugin` as an addon for [Tackle2](https://github.com/konveyor?q=tackle2&type=all&language=&sort=), enabling Java applications with Maven integration to have container image and Kubernetes manifests built for them.
```

File: README.md [hunk 3]
```diff
@@ -1,7 +1,103 @@
...
 
-This addon generate Kubernetes resources from a Java application.
-It uses the [JKube](https://www.eclipse.org/jkube/docs/kubernetes-maven-plugin) Maven plugin to generate the Kubernetes resources.
+This is an implementation of [Jkube's](https://www.eclipse.org/jkube) `kubernetes-maven-plugin` as a konveyor addon. This addon is a small step towards Application Modernization. It brings your <strong>Java</strong> applications on to Kubernetes by leveraging the tasks required to make your application cloud-native. It provides a tight integration into Maven and benefits from the build configuration already provided. 
 
+This addon focus on two tasks: 
+- Generates container images with flexible and powerful configuration.
+- Generates vanilla Kubernetes descriptors (YAML files).
+
+It leverage the [Kubernetes Maven plugin](https://www.eclipse.org/jkube/docs/kubernetes-maven-plugin)  to generate conatiner image and Kubernetes resources.
```
Comment (djzager) [inline README.md]:
```suggestion
It leverage the [Kubernetes Maven plugin](https://www.eclipse.org/jkube/docs/kubernetes-maven-plugin)  to generate conatiner image and Kubernetes manifests.
```

File: README.md [hunk 4]
```diff
@@ -1,7 +1,103 @@
...
+    "name":"Jkube",
+    "state": "Ready",
+    "locator": "jkube",
+    "addon": "jkube",
+    "application": {"id": 2},
+    "data": {}
+}')"
+```
+
+In above request we mention `addon: jkube` as an addon identifier and `application: {id: 2}` states the application against which this addon will be executed. The `data` field here is empty but we can pass data to addon during api request and we use following struct to receive data for the addon.
```
Comment (djzager) [inline README.md]: Should find a way to include actually creating this application, either via the UI or `curl`.

File: README.md [hunk 5]
```diff
@@ -1,7 +1,103 @@
...
+	// Sources list.
+	Sources Sources `json:"sources"`
+	// Targets list.
+	Targets Targets `json:"targets"`
+	// Scope options.
+	Scope Scope `json:"scope"`
+	// Rules options.
+	Rules *Rules `json:"rules"`
+}
+```
```
Comment (djzager) [inline README.md]: I think you could just include a link to the windup addon's Data struct as a reference.

File: README.md (hunk 4 above)
Comment (djzager) [reply README.md]: The `data` field here is empty because the JKube addon accepts none? Please be more specific here.

File: README.md [hunk 6]
```diff
@@ -1,7 +1,103 @@
...
+
+Once a request is received - addon will fetch the application mentioned in the request and perform the operations defined in the addon. For this addon we perform following operations:
+- Fetch the application (clone github/subversion repository)
+- Parse the maven config file and add kubernetes-maven-plugin to `pom.xml`
```
Comment (djzager) [inline README.md:38]: I see now that the windup addon uses maven integration from [here](https://github.com/konveyor/tackle2-addon/blob/main/repository/maven.go), might want to create an issue against this project to explore the maven integration if it buys us anything.

File: README.md [hunk 7]
```diff
@@ -1,7 +1,103 @@
...
+- Build the application using maven build
+- Use kubernetes-maven-plugin to generate container image and k8s manifests
+- Commit the generated resources to repo
+
+## Development Environment Setup
+
+Steps to successfully execute Jkube addon:
+- Setup Minikube and install tackle - [start-minikube action](https://github.com/konveyor/tackle2-operator/tree/main/.github/actions/start-minikube), [install-tackle action](https://github.com/konveyor/tackle2-operator/tree/main/.github/actions/install-tackle), [install-tackle](https://github.com/konveyor/tackle2-operator/blob/main/hack/install-tackle.sh)
```
Comment (djzager) [inline README.md]: You should be able to say something like:
1. Clone this repo
2. Run `make start-minikube install-tackle`.

You can talk about what happens in those two targets and you can reference the actions if you would like.

File: README.md [hunk 8]
```diff
@@ -1,7 +1,103 @@
...
+
+- Forward port for tackle-ui service
+  ```bash
+  kubectl port-forward service/tackle-ui 8080:8080 -n konveyor-tackle
+  ```
+- Populate required data in tackle, including application against which addon will be executed - [reference](https://github.com/konveyor/tackle2-hub/tree/main/hack/add), [jkube-data](https://github.com/konveyor/tackle2-addon-jkube/blob/main/hack/task.sh)
+
+- Add Jkube addon to tackle - [jkube addon](https://github.com/konveyor/tackle2-addon-jkube/blob/main/hack/addon.yml)
+  ```bash
+  kubectly apply -f addon.yml
```
Comment (djzager) [inline README.md]: probably easier to just use the path relative to the project root.

File: README.md [hunk 9]
```diff
@@ -1,7 +1,103 @@
...
+  ```
+- Make curl request to Jkube addon - [request]()
```
Comment (djzager) [inline README.md]: WDYT?
```suggestion
- Create task to run JKube addon against our application
```

File: README.md [hunk 10]
```diff
@@ -1,7 +1,103 @@
...
+  host="${HOST:-localhost:8080/hub}"
+
+  # Make a request to hub
+  request_cmd="$(curl -i -o - -X POST ${host}/tasks -d \
+  '{
+      "name":"Jkube",
+      "state": "Ready",
+      "locator": "jkube",
+      "addon": "jkube",
+      "application": {"id": 1},
```
Comment (djzager) [inline README.md]: You'll need to tell people how to create an application, either through the UI or command line.

File: README.md [hunk 11]
```diff
@@ -1,7 +1,103 @@
...
+      "data": {}
+  }')"
+  ```
+  
+<strong>Reference - </strong> End-to-End setup to run an addon - [Test Windup Addon](https://github.com/konveyor/tackle2-addon-windup/blob/84174448d3d7cd2abc7ba6ab27e66a55890b9061/.github/workflows/test-windup.yml#L36-L53)
```
Comment (djzager) [inline README.md]: What are we trying to communicate with this?

File: README.md [hunk 12]
```diff
@@ -1,7 +1,103 @@
...
+<strong>Reference - </strong> End-to-End setup to run an addon - [Test Windup Addon](https://github.com/konveyor/tackle2-addon-windup/blob/84174448d3d7cd2abc7ba6ab27e66a55890b9061/.github/workflows/test-windup.yml#L36-L53)  
+  
+ ## Output
+* Deployment artifacts
+  * Dockerfile
+  * Kubernetes Yamls
+  
+# Jkube Addon Demonstration
+
+https://user-images.githubusercontent.com/20452032/201997511-3b2a86cb-dfd0-4a0a-9c90-bbc0fdbd987e.mov
```
Comment (djzager) [inline README.md:106]: Can we upload this to YouTube on Konveyor account? cc @savitharaghunathan

**Review by djzager — COMMENTED** [review]
Great start.

File: README.md (hunk 12 above)
Comment (savitharaghunathan) [reply README.md:106]: yes. I will upload it right away and paste the short url here

Comment (savitharaghunathan) [reply README.md:106]: its publishing now and should be available at https://youtu.be/fJM10cq7txg

File: README.md (hunk 4 above)
Comment (mundra-ankur) [reply README.md]: I've shown in the demo, how we can create an application
Should I mention all the steps needed to create an application here?

Comment (mundra-ankur) [reply README.md]: Yeah I might have missed that, in Jkube we aren't accepting any data and that's why it's empty

File: README.md (hunk 11 above)
Comment (mundra-ankur) [reply README.md]: It was supposed to communicate what is required to test an addon and how we are doing it using GitHub actions.

File: README.md (hunk 10 above)
Comment (mundra-ankur) [reply README.md]: I did mention something about that on just two steps above this:

- Populate required data in tackle, including application against which addon will be executed - [reference](https://github.com/konveyor/tackle2-hub/tree/main/hack/add), [jkube-data](https://github.com/konveyor/tackle2-addon-jkube/blob/main/hack/task.sh)

File: README.md (hunk 4 above)
Comment (djzager) [reply README.md]: I believe you should also explain in text here how to create an application, yes.

File: README.md (hunk 10 above)
Comment (djzager) [reply README.md]: I see. Maybe better to be more explicit? ie. Create an application in tackle with this command (and that application is one jkube addon will work correctly). We should also explain so long as we are using `curl` how to get the application's id for use here.

File: README.md (hunk 11 above)
Comment (djzager) [reply README.md]: Ahh great idea. Could you give it it's own section and expand on it please?

File: Makefile [hunk 13]
```diff
@@ -9,4 +9,30 @@ vet:
 	go vet ./...
 
 cmd: fmt vet
-	go build -ldflags="-w -s" -o bin/addon github.com/mundra-ankur/tackle2-addon-jkube/cmd
\ No newline at end of file
+	go build -ldflags="-w -s" -o bin/addon github.com/mundra-ankur/tackle2-addon-jkube/cmd
```
Comment (djzager) [inline Makefile]:
```suggestion
	go build -ldflags="-w -s" -o bin/addon github.com/konveyor/tackle2-addon-jkube/cmd
```

File: README.md (hunk 6 above)
Comment (djzager) [reply README.md:38]: Please create this issue.

**Review by djzager — APPROVED** [review]
Couple comments. Otherwise looks good to me.

File: README.md (hunk 6 above)
Comment (mundra-ankur) [reply README.md:38]: https://github.com/konveyor/tackle2-addon/issues/24

**Pull request merged**